File: 01_Two_sum/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        /*for (int i = 0; i < nums.size(); i++) {
            for (int j = i+1; j < nums.size(); j++) {
                if (nums[i] + nums[j] == target) {
                    return {i,j};
                }
            }
        }

        return {};*/

        // Another solution: more optimal 
        // store {number_value, its index}
        std::unordered_map<int, int> seen;

        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];

            if (seen.count(complement)) {
                return {seen[complement], i};
            }

            seen[nums[i]] = i;
        }
        return {};
    }
};

File: 01_Two_sum/solution.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        seen = {}
        
        for i, num in enumerate(nums):
            complement = target - num

            if complement in seen:
                # return the index of that matching value in the dictionary
                return [seen[complement], i]

            # add the current number and its index to the dictionary
            seen[num] = i

        return []
        